core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
```

Comment:
This looks a lot like the `DataInput` interface in java.

Reply:
Funny that :)  I looked at `DataInput` and our exiting code in htsjdk, and came to the opinion that it would be better to have our own interface that we can evolve as we see fit without being constrained to some existing type.  Also every method in `DataInput` throws `IOException` which means `try / catch` all over the place.  If we're going to use unchecked exceptions that would mean all clients of `ByteSource` have to do exception handling, as opposed to dealing with it at the lower level.